src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,222 @@
+from collections import defaultdict
+from functools import reduce
+from operator import or_
+from typing import List, Mapping, MutableMapping, Optional, Sequence, Set, Union
+
+from django.db.models import Q
+
+from sentry.sentry_metrics.indexer.models import StringIndexer as StringIndexerTable
+from sentry.sentry_metrics.indexer.models import indexer_cache
+from sentry.utils import metrics
+from sentry.utils.services import Service
+
+_INDEXER_CACHE_METRIC = "sentry_metrics.indexer.memcache"
+
+
+class PGStringIndexerV2(Service):
+    """
+    Provides integer IDs for metric names, tag keys and tag values
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _record_many_strings(self, keys: Sequence[str]) -> Mapping[str, int]:
+        """
+        Bulk create db records for org_id:string pairs. As noted below, bulk_create with
+        ignore_conflicts=True returns objects that don't have the pk set on them (aka the `id`
+        that we need), so that's why we use `_get_db_ids_and_set_cache` to look up the records
+        and return a mapping of org_id:string -> id:
+            {
+                "1:release": 3,
+                "2:v1": 4
+            }
+        """
+        records = []
+        for key in keys:
+            organization_id, string = key.split(":")
+            records.append(StringIndexerTable(organization_id=int(organization_id), string=string))
+
+        with metrics.timer("sentry_metrics.indexer.pg_bulk_create"):
+            # We use `ignore_conflicts=True` here to avoid race conditions where metric indexer
+            # records might have be created between when we queried in `bulk_record` and the
+            # attempt to create the rows down below.
+            StringIndexerTable.objects.bulk_create(records, ignore_conflicts=True)
+
+        # We re-query the database to fetch the rows (which should all exist point at this point),
+        # in addition to setting the results in the cache
+        results = self._get_db_ids_and_set_cache(keys)
+        # we _just_ created this records, there is no reason for results to be None
+        assert results
+        return results
+
+    def _get_db_ids_and_set_cache(self, keys: Sequence[str]) -> Union[Mapping[str, int], None]:
```

Comment:
This one as well would be more readable if it took `Tuple[int, str]`
Also when would you return None ? If there are no results why not just returning an empty dict ? So the call site does not have care of dealing with two cases that should mean the same thing.

Also `Union[Mapping[str, int]]` is `Optional[Mapping[str, int]]`